libminifi/src/utils/tls/ExtendedKeyUsage.cpp

```diff
@@ -0,0 +1,104 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifdef OPENSSL_SUPPORT
+
+#include "utils/tls/ExtendedKeyUsage.h"
+
+#include <openssl/x509v3.h>
+
+#include <array>
+#include <cassert>
+#include <climits>
+
+#include "core/logging/LoggerConfiguration.h"
+#include "utils/StringUtils.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+namespace tls {
+
+namespace {
+
+struct KeyValuePair {
+  const char* key;
+  uint8_t value;
+};
+constexpr std::array<KeyValuePair, 6> EXT_KEY_USAGE_NAME_TO_BIT_POS{{
+    KeyValuePair{"Server Authentication", 1},
+    KeyValuePair{"Client Authentication", 2},
+    KeyValuePair{"Code Signing", 3},
+    KeyValuePair{"Secure Email", 4},
+    KeyValuePair{"Time Stamping", 8},
+    KeyValuePair{"OCSP Signing", 9}
+}};
+
+}  // namespace
+
+void EXTENDED_KEY_USAGE_deleter::operator()(EXTENDED_KEY_USAGE* key_usage) const { EXTENDED_KEY_USAGE_free(key_usage); }
+
+ExtendedKeyUsage::ExtendedKeyUsage() : logger_(core::logging::LoggerFactory<ExtendedKeyUsage>::getLogger()) {}
+
+ExtendedKeyUsage::ExtendedKeyUsage(const EXTENDED_KEY_USAGE& key_usage_asn1) : ExtendedKeyUsage{} {
+  const int num_oids = sk_ASN1_OBJECT_num(&key_usage_asn1);
+  for (int i = 0; i < num_oids; ++i) {
+    const ASN1_OBJECT* const oid = sk_ASN1_OBJECT_value(&key_usage_asn1, i);
+    assert(oid && oid->length > 0);
+    const unsigned char bit_pos = oid->data[oid->length - 1];
+    if (bit_pos < CHAR_BIT * sizeof(bits_)) {
+      bits_ |= (1 << bit_pos);
+    }
+  }
+}
+
+ExtendedKeyUsage::ExtendedKeyUsage(const std::string& key_usage_str) : ExtendedKeyUsage{} {
+  const std::vector<std::string> key_usages = utils::StringUtils::split(key_usage_str, ",");
+  for (const auto& key_usage : key_usages) {
+    const std::string key_usage_trimmed = utils::StringUtils::trim(key_usage);
+    const auto it = std::find_if(EXT_KEY_USAGE_NAME_TO_BIT_POS.begin(), EXT_KEY_USAGE_NAME_TO_BIT_POS.end(),
+                                 [key_usage_trimmed](const KeyValuePair& kv){ return kv.key == key_usage_trimmed; });
+    if (it != EXT_KEY_USAGE_NAME_TO_BIT_POS.end()) {
+      const uint8_t bit_pos = it->value;
+      bits_ |= (1 << bit_pos);
+    } else {
+      logger_->log_error("Ignoring unrecognized extended key usage type %s", key_usage_trimmed);
+    }
+  }
+}
+
+bool operator<=(const ExtendedKeyUsage& left, const ExtendedKeyUsage& right) {
+  return (left.bits_ & right.bits_) == left.bits_;
+}
```

Comment:
Take `A=1011` and `B=1100`. `A <= B` is false and `B <= A` is false. This should be impossible if this is meant to be used as a comparison operator. If this is meant to be some kind of arrow, then it needs a comment explaining the operation, or even better, a name.

Reply:
I think it is more of a "subset" relationship, agree that the operator is misleading

Reply:
Yes, it is a subset relation, which is a partial ordering.  `<=` seemed to me like a natural notation for this, but apparently it's not natural to anyone else. :)

I have changed it from `<=` to `isSubsetOf()`.